src/interface/update.hpp

```diff
@@ -13,44 +13,94 @@
 #ifndef INTERFACE_UPDATE_HPP_
 #define INTERFACE_UPDATE_HPP_
 
+#include <algorithm>
+#include <functional>
+#include <limits>
 #include <memory>
 #include <string>
+#include <typeinfo>
+#include <utility>
 #include <vector>
 
 #include "defs.hpp"
-#include "interface/mesh_data.hpp"
-#include "interface/meshblock_data.hpp"
-#include "mesh/mesh.hpp"
+#include "interface/metadata.hpp"
+#include "interface/params.hpp"
+#include "interface/state_descriptor.hpp"
 
 #include "kokkos_abstraction.hpp"
 
 namespace parthenon {
 
 namespace Update {
 
-TaskStatus FluxDivergenceBlock(std::shared_ptr<MeshBlockData<Real>> &in,
-                               std::shared_ptr<MeshBlockData<Real>> &dudt_cont);
-TaskStatus FluxDivergenceMesh(std::shared_ptr<MeshData<Real>> &in_pack,
-                              std::shared_ptr<MeshData<Real>> &dudt_pack);
-void UpdateMeshBlockData(std::shared_ptr<MeshBlockData<Real>> &in,
-                         std::shared_ptr<MeshBlockData<Real>> &dudt_cont, const Real dt,
-                         std::shared_ptr<MeshBlockData<Real>> &out);
-void UpdateMeshData(std::shared_ptr<MeshData<Real>> &in,
-                    std::shared_ptr<MeshData<Real>> &dudt, const Real dt,
-                    std::shared_ptr<MeshData<Real>> &out);
-void AverageMeshData(std::shared_ptr<MeshData<Real>> &c1_pack,
-                     std::shared_ptr<MeshData<Real>> &c2_pack, const Real wgt1);
-Real EstimateTimestep(std::shared_ptr<MeshBlockData<Real>> &rc);
+template <typename T>
+TaskStatus FluxDivergence(std::shared_ptr<T> &in, std::shared_ptr<T> &dudt_obj);
```

Comment:
As mentioned before, I'd rather these go back into the implementation file with the specializations explicitly declared in the `cpp` file. But it's not a huge deal either way.

Reply:
For these sorts of short templated functions, I actually prefer they stay where they are.  Putting them in the cpp file will just lead to more SLOC and a slightly less flexible implementation (e.g. what if we change `MeshBlockData` to `SomeNewShinyName`?  This would be just one more place we'd have to touch.).  I'm going to leave them where they are unless there's more push back.

Reply:
The reason I prefer splitting prototypes and implementations, even for short templated functions, is I think it's easier to read what functions are available and what they do.

That said, I don't care that much.